tlv/record_type.go

```diff
@@ -0,0 +1,59 @@
+package tlv
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"golang.org/x/exp/constraints"
+)
+
+// Primitive is a type constraint that capture the set of "primitive" types,
+// which are the built in stdlib types, and type defs of those types
+type Primitive interface {
+	constraints.Integer | ~[]byte | ~[32]byte | ~[33]byte | ~bool |
+		~*btcec.PublicKey | ~[64]byte
+}
+
+// RecordT is a high-order type makes it easy to encode known "primitive" types
+// as TLV records.
+type RecordT[T Record | Primitive] struct {
+	// Val is the value of the underlying record. Go doesn't let us just
+	// embed the type param as a struct field, so we need an intermediate
+	// variable.
+	Val T
+
+	// Type is the type of the TLV record.
+	Type Type
+}
+
+// NewPrimitiveRecord creates a new RecordT type from a given primitive type.
+func NewPrimitiveRecord[T Primitive](val T, tlvType Type) RecordT[T] {
+	return RecordT[T]{
+		Val:  val,
+		Type: tlvType,
+	}
+}
+
+// NewRecordT creates a new RecordT type from a given Record type. This is
+// useful to wrap a given record in this utility type, which also serves as an
+// extra type annotation. The underlying type of the record is retained.
+func NewRecordT[T Record](record T) RecordT[T] {
```

Comment:
This is not a generic function if I understand go generics correctly. The only valid type here is `Record` no? Or does this mean that anything that implements all of the methods for which `Record` and `*Record` is a receiver then it is implicitly a subtype of `Record` and can be used?

Reply:
This'll fail at compile time if you try to call it with anything that isn't already a `Record`.

Reply:
Ah, I realize now I ran into this: https://github.com/golang/go/issues/49602

For now, the interfaces used in the constraint union can't actually have methods... 

What we actually want here though is `RecordProducer`, but then things don't compile as is since types are being mixed. So we could split it into two types, one that'll save the boiler plate of having to make a new type for a primitive type, and one that'll let you reuse an existing type with a `RecordProducer` implementation in a type declaration.

Reply:
My question here is why isn't the signature like this:

```go
func NewRecordT(record Record) RecordT[Record] { ... }
```